File: services/tradelog/tradelog-api/src/main/java/com/example/tradelog/api/converter/protobuf/TransactionModelConverter.java
package com.example.tradelog.api.converter.protobuf;

import com.example.common.converter.TimeConverter;

import java.util.function.Function;

public class TransactionModelConverter {
    public static Function<com.example.tradelog.api.proto3.model.TransactionModel, com.example.tradelog.api.core.model.TransactionModel> toModel = protoModel -> {
        return new com.example.tradelog.api.core.model.TransactionModel(
                protoModel.getId(),
                protoModel.getAccountId(),
                TimeConverter.toOffsetDateTime.apply((protoModel.getDate())),
                protoModel.getSymbol(),
                TransactionTypeConverter.toModel.apply(protoModel.getType()),
                protoModel.getBrokerFees(),
                TransactionSettinsModelConverter.toModel.apply(protoModel.getSettings())
        );
    };

    public static Function<com.example.tradelog.api.core.model.TransactionModel, com.example.tradelog.api.proto3.model.TransactionModel> toProto = apiModel -> {
        return com.example.tradelog.api.proto3.model.TransactionModel.newBuilder()
                .setId(apiModel.getId())
                .setAccountId(apiModel.getAccountId())
                .setDate(apiModel.getDate().toString())
                .setSymbol(apiModel.getSymbol())
                .setType(TransactionTypeConverter.toProto.apply(apiModel.getType()))
                .setBrokerFees(apiModel.getBrokerFees())
                .setSettings(TransactionSettinsModelConverter.toProto.apply(apiModel.getSettings()))
                .build();
    };
}

File: services/tradelog/tradelog-api/src/main/kotlin/com/example/tradelog/api/rest/converter/TransactionModelConverter.kt
package com.example.tradelog.api.rest.converter

import com.example.common.converter.TimeConverter
import com.example.tradelog.api.core.model.TransactionModel
import com.example.tradelog.api.core.model.TransactionType
import com.example.tradelog.api.spec.model.TLTransactionDto
import java.util.*

class TransactionModelConverter {

    companion object {

        fun toModel(dto: TLTransactionDto): TransactionModel {
            return TransactionModel(
                    id = if (dto.id == null || dto.id.isEmpty() || "null" == dto.id) { UUID.randomUUID() } else { UUID.fromString(dto.id) },
                    accountId = UUID.fromString(dto.accountId),
                    portfolioId = UUID.fromString(dto.portfolioId),
                    date = TimeConverter.toOffsetDateTime(dto.date),
                    symbol = dto.symbol,
                    type = TransactionType.lookup(dto.type.name),
                    brokerFees = dto.brokerFees,
                    settings = TransactionSettingsModelConverter.toModel(dto.settings)
            )
        }

        fun toDto(model: TransactionModel): TLTransactionDto {
            return TLTransactionDto.newBuilder()
                    .setId(model.id.toString())
                    .setAccountId(model.accountId.toString())
                    .setPortfolioId(model.portfolioId.toString())
                    .setDate(model.date.toString())
                    .setSymbol(model.symbol)
                    .setType(TLTransactionDto.TransactionType.valueOf(model.type.name))
                    .setBrokerFees(model.brokerFees)
                    .setSettings(TransactionSettingsModelConverter.toDto(model.settings))
                    .build()
        }
    }
}
